src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCache.scala

```diff
@@ -79,20 +85,37 @@ class ClusterDnsCache(proxyConfig: ProxyConfig, dbRef: DbReference) extends Acto
     }.map(ProcessClusters.apply)
   }
 
+  private def host(c: Cluster): Host = Host(c.googleId.toString + proxyConfig.jupyterDomain)
+
+  private def hostToIpEntry(c: Cluster): (Host, IP) = host(c) -> c.hostIp.get
+
+  private def projectNameToHostEntry(c: Cluster): ((GoogleProject, ClusterName), GetClusterResponse) = {
+    if (c.hostIp.isDefined)
+      (c.googleProject, c.clusterName) -> ClusterReady(host(c))
+    else
+      (c.googleProject, c.clusterName) -> ClusterNotReady
+  }
+
   def processClusters(clusters: Seq[Cluster]): Unit = {
     // Only populate the HostToIp map for clusters with an IP address
     val clustersWithIp = clusters.filter(_.hostIp.isDefined)
-    ClusterDnsCache.HostToIp = clustersWithIp.map(c => Host(c.googleId.toString + proxyConfig.jupyterDomain) -> c.hostIp.get).toMap
+    ClusterDnsCache.HostToIp = clustersWithIp.map(hostToIpEntry).toMap
 
     // Populate the ProjectNameToHost map with all clusters
-    ProjectNameToHost = clusters.map {
-      case c if c.hostIp.isDefined =>
-        (c.googleProject, c.clusterName) -> ClusterReady(Host(c.googleId.toString + proxyConfig.jupyterDomain))
-      case c =>
-        (c.googleProject, c.clusterName) -> ClusterNotReady
-    }.toMap
+    ProjectNameToHost = clusters.map(projectNameToHostEntry).toMap
 
     logger.debug(s"Saved ${clusters.size} clusters to DNS cache, ${clustersWithIp.size} with IPs")
   }
 
+  def processReadyCluster(cluster: Cluster): Future[GetClusterResponse] = {
+    if (cluster.hostIp.isEmpty)
+      Future.failed(ClusterNotReadyException(cluster.googleProject, cluster.clusterName))
+    else Future {
```

Comment:
do you actually want this block running async? if it's all nonblocking serial work you can just have this as a normal `else` block and return the last line in `Future.successful()`

Reply:
It doesn't have to be async but it does need to have error vs. success semantics.  So it could be `Try` or `Either` - except that the result does have to be a `Future` to play nice with actors.

Feel free to change it.

Reply:
👎  on modifying actor state inside a `Future`. I think I would vote for a `Try` or `Either` instead. The actor `ask` call should return a future (see comment below).

Reply:
oh right, good call

Reply:
is there a way to enforce this somehow? i'm glad we're catching it, but having to catch it manually gives me the heebie jeebies

Reply:
It's why FP purists hate akka -- it lets you do untyped stuff and mutable state "safely" without compiler help.

Reply:
(I don't know of a way to enforce it -- it'd be cool if there was some compile time warning or something)

Reply:
Oh yuck.  Thanks.